ADRs/0011-Python-Path-Resolution.md

```diff
@@ -0,0 +1,322 @@
+
+-   id: 2
+    location: C:\Users\shams\miniconda2\Scripts\conda.exe
+    version: 4.7.1
+    ----------------------------------------ENVIRONMENTS---------------------------------------------
+    -   name: base
+        location: C:\Users\shams\miniconda2
+        version: 2.6.1
+    
+    -   name: py27
+        location: C:\Users\shams\miniconda2\envs\py27
+        version: 2.7.7
+    
+    -   name: py25
+        location: C:\Users\shams\miniconda2\envs\py25
+        version: 2.5.1
+    -------------------------------------------------------------------------------------------------
+-----------------------------------------------------------------------------------------------------
+```
+
+Based on the above info (gained from the command: `konduit pythonpaths`), it's much easier to specify python paths in configuration files (using `type`, `location`, `environment` properties):
+
+```json
+{
+  "@type" : "PYTHON",
+  "pythonConfig" : {
+    
+    "type": "conda",
+    "location": "1",
+    "environment": "py37", 
+
+    "python" : "run.py",
+    "importCodePath" : "imports.py",
+    "pythonInputs" : { },
+    "pythonOutputs" : { },
+    "extraInputs" : { },
+    "returnAllInputs" : false,
+    "setupAndRun" : false,
+    "ioInputs" : {
+      "images" : {
+        "name" : "images",
+        "pythonType" : "list",
+        "secondaryType" : "NDARRAY",
+        "type" : "LIST"
+      }
+    },
+    "ioOutputs" : {
+      "texts" : {
+        "name" : "texts",
+        "pythonType" : "list",
+        "secondaryType" : "STRING",
+        "type" : "LIST"
+      },
+      "boxes" : {
+        "name" : "boxes",
+        "pythonType" : "list",
+        "secondaryType" : "NDARRAY",
+        "type" : "LIST"
+      }
+    },
+    "jobSuffix" : "konduit_job"
+  }
+}
+```
+
+#### Possible Values
+- The `type` property will take one of the following values: [`python`, `conda`, `custom`]
+- `location` will have the *id* of the python install for `type==python` or the id of the conda install for `type==conda`. Otherwise, it will contain the absolute location of the python executable file when `type==custom`
+- `environment` will be looked into if `type==conda` to locate the python installation for a particular conda environment.
+                               
+### Defaults and Priorities
+The defaults for python paths are to be expected in the following way:
+
+![Python Paths Defaults](../markdown_images/python-installation-paths-defaults.png)
+
+### Usage with Profiles
+The profiles should be extended to incorporate the python paths settings for running with the `serve` command. The same items names that are used for the PythonStep config are going to be used here. For example: 
+
+```shell script
+konduit profile create python_37_env --python-type=conda --python-location=1 --conda-env=py37 <...other_options...>
+                                        # OR
+konduit profile create python_37_env -pt=conda -pl=1 -ce=py37 <...other_options...>
+```
```

Comment:
Another note might be virtual envs as well. I'm not sure how hard it is, but specifying whether we create a separate virtual env could also be good. Same for running conda. I wonder if we could bootstrap a conda environment from   konduit as well. As long as we have access to a python binary we're good to go.

Reply:
Hmm. Virtualenv would more or less have the same workflow so shouldn't be hard to add it here. We basically have a lot of options to do here but most of it is running conda/python/virtualenv commands. There's no problem using the basic few commands for listing packages or versions but using dependencies resolution for python would take a bit more effort here. It's fine for cases where there are no clashes between deps but resolving clashes is gonna be a mess if we try to do it directly from konduit.

Reply:
I'll keep bootstrapping new python environments (whether `virtualenv` or `conda`) for a separate PR. For this ADR, I'll just work on specifying pre-installed/created python installs/environments.